src/resources/spi_linux.cc

```diff
@@ -26,75 +27,242 @@
+    USE(resource);
+    state |= data;
+    return state;
+  }
+};
+
+
+class SpiResource : public Resource {
+ public:
+  TAG(SpiResource);
+  SpiResource(ResourceGroup* group);
+  ~SpiResource();
+
+  int fd() { return fd_; }
+  void set_fd(int fd) {
+    ASSERT(fd_ == -1);
+    fd_ = fd;
+  }
+
+  int error() const { return error_; }
+  void set_error(int error) { error_ = error; }
+
+  uint8* buffer() { return buffer_; }
+
+  Object* transfer_start(Blob data, int from, int length,
+                         bool is_read,
+                         int delay_usecs,
+                         bool cs_change,
+                         Process* process);
+
+  Object* transfer_finish(bool was_read, Process* process);
+
+ private:
+  int fd_ = -1;
+  int error_ = 0;
+  AsyncEventThread* thread_ = null;
+  int buffer_size_ = 0;
+  uint8* buffer_ = null;
+};
+
+void SpiResourceGroup::on_unregister_resource(Resource* resource) {
+  int fd = static_cast<SpiResource*>(resource)->fd();
+  int result = EINTR;
+  while (result == EINTR) {
+    result = close(fd);
+  }
+}
+
+SpiResource::SpiResource(ResourceGroup* group) : Resource(group) {}
+
+SpiResource::~SpiResource() {
+  if (thread_ != null) {
+    delete thread_;
+  }
+  if (fd_ >= 0) {
+    close(fd_);
+  }
+  if (buffer_ != null) {
+    free(buffer_);
+  }
+}
+
+Object* SpiResource::transfer_start(Blob data, int from, int length,
+                                    bool is_read,
+                                    int delay_usecs,
+                                    bool cs_change,
+                                    Process* process) {
+  bool successfully_dispatched = false;
+  if (buffer_ != null) FAIL(INVALID_STATE);
+  if (length <= 0 || delay_usecs < 0 || delay_usecs > 0xffff) FAIL(OUT_OF_BOUNDS);
+  if (from < 0 || from + length > data.length()) FAIL(OUT_OF_BOUNDS);
+
+  // Since we are returning to the user, we can't hold onto the data and need to copy it.
+  // TODO(florian): allow to neuter incoming external byte arrays.
+  uint8* buffer = unvoid_cast<uint8*>(malloc(length));
+  if (buffer == null) FAIL(MALLOC_FAILED);
+  Defer free_buffer{ [&] { if (!successfully_dispatched) free(buffer); } };
+
+  memcpy(buffer, data.address() + from, length);
+
+  auto tx_address = buffer;
+  // Reuse the same buffer for reading.
+  auto rx_address = is_read ? buffer : null;
+
+  auto xfer = unvoid_cast<struct spi_ioc_transfer*>(calloc(1, sizeof(struct spi_ioc_transfer)));
+  if (xfer == null) FAIL(MALLOC_FAILED);
+  Defer free_xfer{ [&] { if (!successfully_dispatched) free(xfer); } };
+
+  xfer->tx_buf = reinterpret_cast<uint64>(tx_address);
+  xfer->rx_buf = reinterpret_cast<uint64>(rx_address);
+  xfer->len = length;
+  xfer->delay_usecs = delay_usecs;
+  // TODO(florian): this is probably inverted.
+  // See: https://github.com/beagleboard/kernel/issues/85
+  xfer->cs_change = cs_change ? 0x01 : 0x00;
+
+  if (thread_ == null) {
```

Comment:
Still not entirely clear why this cannot be closer to a LazyEventSource. Maybe worth chatting about.

Reply:
The only difference is that the SPI device here doesn't shut down the thread on "stop". Right?

Reply:
as discussed. The shutting down is fine.
the main difference is that each resource needs its own thread.